design-decisions/2018-11-05-Translating-predicates.md

```diff
@@ -0,0 +1,122 @@
+otherwise it is unlikely to have the information required to disprove it.
+
+Problems
+--------
+
+Although matching logic's connectives share the names of the solver's
+connectives, they operate differently in general. We are left with two problems:
+
+1. How will we translate builtin `Bool` and `Int` predicates?
+2. How will we translate arbitrary predicates?
+
+Builtin translation
+-------------------
+
+Hooked symbols applied to the builtin `Bool` and `Int` sorts will be translated
+directly using the corresponding SMT operations.
+
+Builtin domain values (`\dv{Bool}` and `\dv{Int}`) are translated to the
+corresponding literals in the solver.
+
+Variables in the `Bool` and `Int` sorts are translated as variables in the
+solver. (This is allowed because matching logic variables are single-valued.)
+
+Equality defined by Kore `\equals{Bool, I}` and `\equals{Int, I}` is translated
+to the SMT equality operation on `Bool` and `Int` respectively. The result of
+the SMT operation inhabits sort `Bool`, which interacts with our decision below
+to translate predicates in arbitrary matching logic sorts to the SMT `Bool`
+sort. SMT equality corresponds to matching logic equality *only* for functional
+patterns, so we take care to only translate functional patterns in builtin
+sorts.
+
+The matching logic connectives are *not* interpreted as connectives in the
+solver because their interpretation differs. For example, consider that in
+matching logic,
+```
+\and{Bool}(true{}(), false{}()) === \bottom{}()
+```
+whereas in the solver,
+```
+true ∧ false === false.
+```
+
+Arbitrary-sort translation
+--------------------------
+
+Predicates in arbitrary sorts are equal either to `\top{I}()` or `\bottom{I}()`,
+so they can be represented in the solver's `Bool` sort.
+
+`\top{I}()` and `\bottom{I}()` can be interpreted by the solver as `true` and
+`false` respectively.
+
+The matching logic connectives (`\and{I}`, `\or{I}`, `\not{I}`, `\implies{I}`,
+`\iff{I}`) actually correspond to their counterparts in the solver, but only
+operating on predicates. Therefore, these patterns are interpreted directly in
+the solver when their children are predicates.
+
+`\ceil{S, I}` and `\floor{S, I}` cannot be interpreted in the solver because
+their children are not predicates. These patterns are predicates, so they are
+translated as variables, i.e. left uninterpreted by the solver. `\in{S, I}` is
+also translated as a variable. Identical patterns will be translated to the
+same variable, so that the solver can reason over the predicates even if it
+cannot reason under them.
+
+As indicated above, `\equals{Bool, I}` and `\equals{Int, I}` are interpreted as
+the corresponding builtin equalities. This is possible because the predicate in
+`I` is translated to `Bool` in the solver. Equality on predicates (`\equals{I,
+I}`) is translated as `\iff{I}`. Equality on all other sorts is translated to a
+variable, i.e. left uninterpreted as `\ceil{S, I}`.
```

Comment:
I wonder whether, for all other sorts, we could actually interpret equality as equality, and have the operands be variables. This might let the SMT do things like transitivity and substitution.

Reply:
I'm almost certain we can't do that for arbitrary sorts because the patterns might not be functional.